### DISCUSSION TITLE
Integrate files in firmware

### DISCUSSION THREAD

**Discussion by utya** [arendst/Tasmota#18989] [Q&A] [answered]
Hi. I want to ask, can I integrate some files in file system of firmware. For sample display.ini and berry script

**Comment by sfromis** [comment, marked as answer]
Yes, using `custom_files_upload`. Here's an example of how it works:
https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_cenv_sample.ini#L17-L35

Beware that the preloaded file system is only applied when flashing the `.factory` variant of the binary, not when merely upgrading.

With a display.ini + Berry, I'd consider if it would be a good candidate for packing in an [autoconf](https://github.com/tasmota/autoconf/).

**Comment by plu** [comment]
I'm also trying to get this working, without success :(. The output looks like this:

```
Successfully created esp32 image.
esp32_create_combined_bin([".pio/build/tasmota32-txultimate/firmware.bin"], [".pio/build/tasmota32-txultimate/firmware.elf"])
Read partitions from  partitions/esp32_partition_app2880k_fs320k.csv
# Name,     Type,   SubType,   Offset,    Size,   Flags
nvs         data    nvs        0x9000     0x5000
otadata     data    ota        0xe000     0x2000
safeboot    app     factory    0x10000    0xD0000
app0        app     ota_0      0xE0000    0x2D0000
spiffs      data    spiffs     0x3B0000   0x50000
Creating filesystem with content:
/autoexec.be
/txultimate.be
Will download safeboot binary from URL:
http://ota.tasmota.com/tasmota32/release/tasmota32-safeboot.bin
safeboot binary written to variants dir.
    Offset | File
 -  0x1000 | /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/bootloader.bin
 -  0x8000 | /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/partitions.bin
 -  0xe000 | /root/.platformio/packages/framework-arduinoespressif32/tools/partitions/boot_app0.bin
 -  0x10000 | /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/variants/tasmota/tasmota32-safeboot.bin
 - 0xe0000 | /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/firmware.bin
 - 0x3b0000| /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/littlefs.bin
Will use custom upload command for flashing operation to add file system defined for this build target.
esptool.py v4.7.3
SHA digest in image updated
Wrote 0x400000 bytes to file /tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/firmware.factory.bin, ready to flash to offset 0x0
bin_map_copy([".pio/build/tasmota32-txultimate/firmware.bin"], [".pio/build/tasmota32-txultimate/firmware.elf"])
map_gzip([".pio/build/tasmota32-txultimate/firmware.bin"], [".pio/build/tasmota32-txultimate/firmware.elf"])
firm_metrics([".pio/build/tasmota32-txultimate/firmware.bin"], [".pio/build/tasmota32-txultimate/firmware.elf"])
/opt/hostedtoolcache/Python/3.12.4/x64/bin/python -m tasmota_metrics "/tmp/runner/gha-runner-1/home-tasmota/home-tasmota/.pio/build/tasmota32-txultimate/firmware.map"
Total sizes:
Used static DRAM:   79000 bytes (  45580 remain, 63.4% used)
      .data size:   27384 bytes
      .bss  size:   51616 bytes
Used static IRAM:   92118 bytes (  79914 remain, 53.5% used)
      .text size:   91091 bytes
   .vectors size:    1027 bytes
Used Flash size : 2104271 bytes
           .text: 1360155 bytes
         .rodata:  743860 bytes
Total image size: 2223773 bytes (.bin may be padded larger)
======================== [SUCCESS] Took 302.76 seconds ========================

Environment           Status    Duration
--------------------  --------  ------------
tasmota32-txultimate  SUCCESS   00:05:02.761
```

So it seems that two files have been added successfully:

```
Creating filesystem with content:
/autoexec.be
/txultimate.be
```

But I cannot see the files on the device, after flashing the firmware (OTA) onto it.

My configuration looks like this:

```
[env:tasmota32-txultimate]
extends = env:tasmota32_base
board_build.f_cpu = 240000000L
build_flags = ${env:tasmota32_base.build_flags} -DFIRMWARE_TXULTIMATE -DFIRMWARE_TASMOTA32
custom_files_upload = ${env:tasmota32_base.custom_files_upload}
                      home-txultimate/autoexec.be
                      home-txultimate/txultimate.be
```

I'd be grateful for any hints on what I might be missing or doing wrong.

**Reply by plu** [reply]
Btw, I just noticed this:

> Beware that the preloaded file system is only applied when flashing the .factory variant of the binary, not when merely upgrading.

That one is quite huge:

```
tasmota32-txultimate.bin                           25-Jun-2024 13:33             2230224
tasmota32-txultimate.factory.bin                   25-Jun-2024 13:33             4194304
```

I cannot flash that one via OTA, right?

I'd like to get files over the air into an image, is there any other way than including them in the image?

**Reply by sfromis** [reply]
Right, the files are indeed not included with OTA image upgrading, as it is to construct a preloaded file system, not as individual files.

If you want to update files OTA after initial flashing, you'll have to transfer individual files. Of course, this does not get them "into the image", only into the file system on the device.

I'm frequently uploading files, like Berry code, to devices, as a much more convenient operation than trying to have them part of images for flashing. For "fixed" needs, I'd be inclined to only have a small uploader in the image, able to fetch the needed files, if not already in the file system.

**Reply by sfromis** [reply]
The files you're using would be good candidates for having in a `.tapp` file or even an `.autoconf` setup. The latter would not be well-integrated with the web UI for upgrading autoconf files.

**Reply by plu** [reply]
Thank you for your answer! Do you have any automation for uploading your files, or are you doing this manually? I'd like to get to a point where I can do a `reset 1` and get any new or old device into the desired state, including configuration and files. For configuration I have a solution already, doing all the configuration via MQTT. But I'm still lacking the files part for the SONOFF TX Ultimate devices.

**Reply by plu** [reply]
Ah! Just found `UrlFetch`, that might do the trick for my usecases!

**Reply by kfalz0n** [reply]
Hi plu, sfromis. I have a question (sorry for jumping in the thread). I am fairly new to flashing tasmota as usually flashed ESPHome but I have bought a couple of TX Ultimate (4 gang) devices and would like to get them with tasmota.

I have been reading the template and whilst I understand the process of flashing the bin and txtultimate.be file I came across needing to do this: "Add a new environment to platformio_override.ini or platformio_tasmota_cenv.ini" which seems relavant to the above. 

How do I add a new environment to the ini file please? Thanks

**Reply by plu** [reply]
In my earlier post you can actually see an example on how to add a new environment:

```
[env:tasmota32-txultimate]
```

This adds a new environment called `tasmota32-txultimate`.

**Reply by kfalz0n** [reply]
Thanks - but did you use esptool.py to send the command to create a new environment? As I was planning on using the Web Installer and not sure if it can be done that way.. so basically to add a new environment I just need to send a command? But how would I send this entire one?

`[env:tasmota32-txultimate]
extends                 = env:tasmota32_base
board_build.f_cpu       = 240000000L
build_flags             = ${env:tasmota32_base.build_flags}
                          -DOTA_URL='"-"'
                          -DCODE_IMAGE_STR='"txultimate"'
                          -DUSE_I2S_AUDIO
                          -DUSE_I2S_LSB ; Audio fix
                          -DUSE_I2S_SAY_TIME
                          -DUSE_I2S_WEBRADIO
                          -DUSE_I2S_RTTTL
`

**Reply by plu** [reply]
`esptool.py` is for flashing images. To build/compile them you'll need [PlatformIO](https://platformio.org).

**Reply by kfalz0n** [reply]
alright I understand now, I will need to build and compile the custom binary and then flash it with the esptool.py to get this new environment embedded. Thank you
